File: 02 - Factory/Python/factory_method.py
from math import sin, cos

class Point:
    def __init__(self, a, b) -> None:
        self.x = a
        self.y = b
    
    def __str__(self) -> str:
        return f'x: {self.x}, y: {self.y}'
    

    @staticmethod
    def cartesian_factory(x, y):
        return Point(x, y)
    
    @staticmethod
    def polar_factory(rho, theta):
        return Point(rho * cos(theta), rho * sin(theta))

if __name__ == "__main__":
    p1 = Point.cartesian_factory(2, 3)
    p2 = Point.polar_factory(2, 3)

    print(p1, p2)

File: 02 - Factory/Dart/factory_method.dart
import 'dart:math' as math;

void main() {
  Point p1 = Point.cartesian(2, 3);
  Point p2 = Point.polar(2, 3);

  print(p1);
  print(p2);
}

class Point {
  final double a;
  final double b;

  const Point(this.a, this.b);

  @override
  String toString() => "x: $a, y:$b";

  factory Point.cartesian(double x, double y) {
    return Point(x, y);
  }

  factory Point.polar(double rho, double theta) {
    return Point(rho * math.cos(theta), rho * math.sin(theta));
  }
}
